curriculum/challenges/english/25-front-end-development/review-react-forms-data-fetching-and-routing/67c9e932c6c4234532d46394.md

```diff
@@ -0,0 +1,426 @@
+---
+id: 67c9e932c6c4234532d46394
+title: React Forms, Data Fetching and Routing Review
+challengeType: 24
+dashedName: review-react-forms-data-fetching-and-routing
+---
+
+# --description--
+
+## Working with Forms in React
+
+- **Controlled Inputs**: This is when you store the input field value in state and update it through `onChange` events. This gives you complete control over the form data and allows instant validation and conditional rendering.
+
+```jsx
+import { useState } from "react";
+
+function App() {
+  const [name, setName] = useState("");
+
+  const handleChange = (e) => {
+    setName(e.target.value);
+  };
+
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    console.log(name);
+  };
+
+  return (
+    <>
+      <form onSubmit={handleSubmit}>
+        <label htmlFor="name">Your name</label> <br />
+        <input value={name} id="name" onChange={handleChange} type="text" />
+        <button type="submit">Submit</button>
+      </form>
+    </>
+  );
+}
+
+export default App;
+```
+
+- **Uncontrolled Inputs**: Instead of handling the inputs through the `useState` hook, uncontrolled inputs in HTML maintain their own internal state with the help of the DOM. Since the DOM controls the input values, what you need is to pull in the values of the input fields with a `ref`. 
```

Comment:
```suggestion
- **Uncontrolled Inputs**: Instead of handling the inputs through the `useState` hook, uncontrolled inputs in HTML maintain their own internal state with the help of the DOM. Since the DOM controls the input values, you need to pull in the values of the input fields with a `ref`. 
```